abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sets the values of the key (`key_data`, `key_size`) in a shared key-value store
+(`kvstore_id`) to the values (`values_data`, `values_size`).
+
+If compare-and-switch value (`cas`) is provided and non-zero, then it must match
+the current value in order for update to succeed.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the key-value store (`kvstore_id`) doesn't exist.
+
+Returns `CompareAndSwapMismatch` when the compare-and-switch value isn't
+current.
+
+Returns `InvalidMemoryAccess` when `key_data`, `key_size`, `values_data` and/or
+`values_size` point to invalid memory address.
+
+#### `proxy_add_shared_kvstore_key_values`
+
+* params:
+  - `i32 (uint32_t) kvstore_id`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (iovec_array_t) values_data`
+  - `i32 (size_t) values_size`
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Adds new values (`values_data`, `values_size`) to the key (`key_data`,
+`key_size`) in a shared key-value store (`kvstore_id`).
+
+If compare-and-switch value (`cas`) is provided and non-zero, then it must match
+the current value in order for update to succeed.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the key-value store (`kvstore_id`) doesn't exist.
+
+Returns `CompareAndSwapMismatch` when the compare-and-switch value isn't
+current.
+
+Returns `InvalidMemoryAccess` when `key_data`, `key_size`, `values_data` and/or
+`values_size` point to invalid memory address.
+
+#### `proxy_remove_shared_kvstore_key`
+
+* params:
+  - `i32 (uint32_t) kvstore_id`
+  - `i32 (const char*) key_data`
+  - `i32 (size_t) key_size`
+  - `i32 (uint32_t) cas`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Removes the key (`key_data`, `key_size`) from a shared key-value store
+(`kvstore_id`).
+
+If compare-and-switch value (`cas`) is provided and non-zero, then it must match
+the current value in order for update to succeed.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the key-value store (`kvstore_id`) doesn't exist.
+
+Returns `CompareAndSwapMismatch` when the compare-and-switch value isn't
+current.
+
+Returns `InvalidMemoryAccess` when `key_data` and/or `key_size` point to invalid
+memory address.
+
+#### `proxy_delete_shared_kvstore`
+
+* params:
+  - `i32 (uint32_t) kvstore_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Deletes previously created shared key-value store (`kvstore_id`).
+
+Returns `Ok` on success.
+
+Returns `BadArgument` when the key-value store (`kvstore_id`) doesn't exist.
+
+
+## Shared queue
+
+Note: Global queue can be always accessed using `queue_id` `0`.
+
+### Functions exposed by the host environment
+
+#### `proxy_open_shared_queue`
+
+* params:
+  - `i32 (const char*) queue_name_data`
+  - `i32 (size_t) queue_name_size`
+  - `i32 (bool) create_if_not_exist`
+  - `i32 (uint32_t*) return_queue_id`
```

Comment:
Are "queue_ids" intended to be unique across all wasm modules on the same host, or only on the same vm, or only on the same module?

I'm asking because in one case I found it useful for a singleton "bootstrap service" to expose a "request queue," and for each instance of another module to create its own "reply queue." The non-singleton module then puts a message on the request queue that includes the queue id of the reply queue, so that the two can implement a request-response workflow.